_posts/2016-09-21-approaching-tdd-outside-android.md

```diff
@@ -0,0 +1,159 @@
+
+<b>Then</b> the statement should be a list with all the transactions in reverse chronological order
+<b>And</b> the statement lines should contain the transaction amount, date and running balance
+
+
+## Setting up the project
+
+Before getting our hands dirty with the code, we have to configure the tools that we are going to use to write our tests. We have chosen <a href="http://junit.org/junit4/">JUnit</a>, <a href="http://mockito.org/">Mockito</a> and <a href="https://google.github.io/android-testing-support-library/docs/espresso/">Espresso</a> for assertions in Android views. We have to add their dependencies to the project build.gradle as follows:
+
+````
+testCompile 'junit:junit:4.12'
+testCompile 'org.mockito:mockito-core:1.10.17'
+
+androidTestCompile "com.android.support.test:runner:0.4.1"
+androidTestCompile "com.android.support.test:rules:0.4.1"
+androidTestCompile "com.android.support.test.espresso:espresso-core:2.2.2"
+androidTestCompile("com.android.support.test.espresso:espresso-contrib:2.2.2") {
+    exclude module: 'recyclerview-v7'
+    exclude module: 'support-v4'
+}
+androidTestCompile 'org.mockito:mockito-core:1.10.17'
+androidTestCompile 'com.google.dexmaker:dexmaker:1.2'
+androidTestCompile 'com.google.dexmaker:dexmaker-mockito:1.2'
+````
+
+And at the bottom of the file:
+
+````
+resolutionStrategy.force "com.android.support:support-annotations:$supportLibraryVersion"
+````
+
+
+For the sake of clarity of this exercise, we are going to use the default source set src/androidTest for our acceptance tests and src/test for the unit tests. This configuration may not be the ideal in a production environment, due to the fact that you would need to add some unit tests for the Android components in your project. If you are in that case, you could end up with a mix of unit/acceptance tests in the same directory, losing the possibility to run all the tests quickly. Remember that acceptance and integration tests are slower than unit tests.
+
+In a real project we will need to run all the unit tests alone to ensure that they are passing and checking the current step in the inner TDD loop. That is why is a good idea to separate them.
+
+## Writing an acceptance test for the acceptance criteria
+
+<img src="http://panavtec.me/wp-content/uploads/2016/04/outside-in.png" alt="outside-in" width="573" height="394" class="aligncenter size-full wp-image-1728" />
+
+This image show the testing flow that we are going to follow: the double loop of TDD. The outside loop corresponds to the progress of our feature and the inner loop corresponds to the individual functionals required to implement the feature.
+
+It's worth defining our test types very clearly.
+<ul>
+<li><b>Unit test</b>: Test that our class does the correct thing</li>
+<li><b>Acceptance test</b>: Test that our system passes the acceptance criteria and therefore behaves properly using real collaborators. Leaving aside external systems such as network, database, API, etc…</li>
```

Comment:
Verify that our system works according to the requirements specified in the acceptance criteria. Here we test the system as a whole, using all the real classes together, but leaving aside external systems like network, database, API, etc.